File: branch_bound_job_assignment.py
import sys

def assign_jobs(cost_matrix, person, current_cost, job_assigned, current_assignment, state, n):
    # Base Case
    if person == n:
        if current_cost < state["min_cost"]:
            state["min_cost"] = current_cost
            state["best_assignment"] = list(current_assignment)
        return

    # Branching & Bounding loop
    for job in range(n):
        if not job_assigned[job]:
            # Bound condition check
            if current_cost + cost_matrix[person][job] < state["min_cost"]:
                
                job_assigned[job] = True
                current_assignment[person] = job

                # Recurse for next person
                assign_jobs(cost_matrix, person + 1, current_cost + cost_matrix[person][job], job_assigned, current_assignment, state, n)

                # Reset state
                job_assigned[job] = False
                current_assignment[person] = -1

def main():
    print("--- DSA Masterclass: Job Assignment using Branch & Bound (Python) ---\n")

    cost_matrix = [
        [9, 2, 7, 8],
        [6, 4, 3, 7],
        [5, 8, 1, 8],
        [7, 6, 9, 4]
    ]
    
    n = len(cost_matrix)
    job_assigned = [False] * n
    current_assignment = [-1] * n
    
    state = {
        "min_cost": sys.maxsize,
        "best_assignment": [-1] * n
    }

    assign_jobs(cost_matrix, 0, 0, job_assigned, current_assignment, state, n)

    print("Optimal Job Allocation Mapping (Python):")
    for i in range(n):
        assigned_job = state["best_assignment"][i]
        print(f"  Person {i} -> assigned to Job {assigned_job} (Cost: {cost_matrix[i][assigned_job]})")
    print("---------------------------------")
    print(f"Minimum Total Safe Cost: {state['min_cost']}\n")

if __name__ == "__main__":
    main()

File: branch_bound_job_assignment.cpp
#include <iostream>
#include <vector>
#include <climits>

using namespace std;

#define N 4 // 4 Persons and 4 Jobs

int minTotalCost = INT_MAX;
vector<int> bestAssignment(N, -1);

// Utility function to find minimum cost allocation using Branch and Bound logic
void assignJobs(vector<vector<int>>& costMatrix, int person, int currentCost, vector<bool>& jobAssigned, vector<int>& currentAssignment) {
    // Base Case: If all persons are assigned a job
    if (person == N) {
        if (currentCost < minTotalCost) {
            minTotalCost = currentCost;
            bestAssignment = currentAssignment;
        }
        return;
    }

    // Branching: Try assigning every available job to the current person
    for (int job = 0; job < N; job++) {
        // If job is not yet assigned
        if (!jobAssigned[job]) {
            
            // BOUND: Skip this branch if the cost already exceeds our minimum found cost
            if (currentCost + costMatrix[person][job] < minTotalCost) {
                
                jobAssigned[job] = true;
                currentAssignment[person] = job;

                // Move to next person
                assignJobs(costMatrix, person + 1, currentCost + costMatrix[person][job], jobAssigned, currentAssignment);

                // Backtrack / Unbind branch
                jobAssigned[job] = false;
                currentAssignment[person] = -1;
            }
        }
    }
}

int main() {
    cout << "--- DSA Masterclass: Job Assignment using Branch & Bound (C++) ---" << endl << endl;

    // Row = Person, Column = Job Cost
    vector<vector<int>> costMatrix = {
        {9, 2, 7, 8},
        {6, 4, 3, 7},
        {5, 8, 1, 8},
        {7, 6, 9, 4}
    };

    vector<bool> jobAssigned(N, false);
    vector<int> currentAssignment(N, -1);

    assignJobs(costMatrix, 0, 0, jobAssigned, currentAssignment);

    cout << "Optimal Job Allocation Mapping:" << endl;
    cout << "---------------------------------" << endl;
    for (int i = 0; i < N; i++) {
        cout << "  Person " << i << " -> assigned to Job " << bestAssignment[i] 
             << " (Cost: " << costMatrix[i][bestAssignment[i]] << ")" << endl;
    }
    cout << "---------------------------------" << endl;
    cout << "Minimum Total Safe Cost: " << minTotalCost << endl;

    return 0;
}